### ISSUE TITLE
make-mode `-M` expects `conf.py` in wrong directory

### ISSUE THREAD

**Issue by vlcinsky** [sphinx-doc/sphinx#4900]
Subject: `-M html` expects `conf.py` in different dir than `-b html`

### Problem
Building e.g. html doc using `-b` and using `-c` to set directory with `conf.py` works well.

However, when using `-M`, the command fails complaining about missing `conf.py` as it apparently searches for it in modified directory (with appended `html`).

#### Procedure to reproduce the problem
Create `spihinx-build` call:
- use `-b` for `html`
- use `-c` to explicitly state, where is `conf.py` directory located
- explicitly set source directory
- set target directory

```
$ /home/javl/devel/vars/formatdoc/.tox/py36/bin/sphinx-build -b html -c /home/javl/devel/vars/formatdoc/test_doc ../../vars-profiles-2/vars-profiles/cz-ndic_d2-common/docs/cs _build/common-cs/html
```
Such command worked for me well.

Then try to change `-b` to `-M`. Theoretically it should behave the same:
```
$ /home/javl/devel/vars/formatdoc/.tox/py36/bin/sphinx-build -b html -c /home/javl/devel/vars/formatdoc/test_doc ../../vars-profiles-2/vars-profiles/cz-ndic_d2-common/docs/cs _build/common-cs/html
usage: sphinx-build [OPTIONS] SOURCEDIR OUTPUTDIR [FILENAMES...]
sphinx-build: error: config directory doesn't contain a conf.py file (/home/javl/devel/vars/formatdoc/test_doc/html)
```
The same applies to `latexpdf` and `latex` builders. It seems that `-M` is not handling location of config directory properly.

#### Expected results
sphinx-build with option -M shall correctly recognize configuration directory set via `-c` option.

### Environment info
- OS: Linux: Debina 9 Stretch
- Python version: 3.6
- Sphinx version: 1.7.1

**Comment by vlcinsky** [comment]
I read the code and tried to figure out, what could go wrong. There seem to be possible reasons:
- `-c` option is ignored with `-M`
- with `-M`, there is some `cd` into subdirectory with name being builder name

**Comment by vlcinsky** [comment]
Reading source code shows, that funciton `run_make_mode`(https://github.com/sphinx-doc/sphinx/blob/master/sphinx/make_mode.py#L157) in module `make_mode`:

``` python
def run_make_mode(args):
    # type: (List[unicode]) -> int
    if len(args) < 3:
        print('Error: at least 3 arguments (builder, source '
              'dir, build dir) are required.', file=sys.stderr)
        return 1
    import pdb; pdb.set_trace()
    make = Make(args[1], args[2], args[3:])
    run_method = 'build_' + args[0]
    if hasattr(make, run_method):
        return getattr(make, run_method)()
    return make.run_generic_build(args[0])
```

does not parse arguments as `sphinx-build -b` variant, assumes fixed positional arguments and ignores whatever comes after `build_dir` argument.

My current problem (configuration directory being expected in wrong directory) was resolved by placing `-c where/conf_py/lives` at the end, as I in my case `conf.py is in current directory and it somehow helps.

Anyway, current `-M` command variant has different argument parsing and seem to ignore many options.

I would propose to change the code to reuse argparse configuration from cmd/build9.py even for `-M` variant.

*EDIT*: `Make` is using all the options, my original comment was wrong in assuming, only first three positional arguments are used.

**Comment by cblegare** [comment]
I had an other issue with the same cause.

Given folder structure:

```
.
├── build
└── doc
    ├── conf.py
    └── index.rst
```

Running from `doc` folder:

```
sphinx-build -M latexpdf -d ../build/sphinx/doctrees -n -T -W -vvv . ../build/sphinx/pdf
```

I put a breakpoint to inspect values in `Make` object in `make_mode.py` and get:

```
builddir='../build/sphinx/doctrees`
makecmd='make'
opts=['-n', '-T', '-W', '-vvv', '.', '../build/sphinx/pdf']
srcdir='-d'
```

Which cause the `cd` in `build_latexpdf` to change dir to `build/sphinx/doctrees` instead of `build/sphinx/pdf` and no makefile is found for the pdflatex build.

Given the documented [`sphinx-build` synopsis](http://www.sphinx-doc.org/en/master/man/sphinx-build.html#synopsis):

```
sphinx-build [options] <sourcedir> <outputdir> [filenames …]
```

---
EDIT

### tl;dr

`sphinx.make_mode` is broken if it can't use `sphinx.cmd.build.get_parser` or equivalent.

### repro steps

Call `sphinx-build -M something` with at least one option. (i.e. `-W`) and the `sphinx.make_mode.Make` class is **not** properly initialized.

### Closer view

Lets simplify `sphinx.make_mode.run_make_mode`

```python
def run_make_mode(args):
    """
   According to the sphinx-build synopsis

    args[0] is always a builder name
    args[1:i] are options
    args[i] is sourcedir
    args[i+1] is outputdir
    args[slice(i+2:)] are excluded filenames
    """
    # find i
    sourcedir = args[i]
    outputdir = args[i+1]
    options = args[1:i] + args[slice(i+2:)]
    make = Make(sourcedir, outputdir, options)
    # do something with make
```

There is also an issue in  `sphinx.make_mode.Make.run_generic_build` regarding [`sphinx-build` synopsis](http://www.sphinx-doc.org/en/master/man/sphinx-build.html#synopsis).

**Comment by AA-Turner** [comment]
`-M` and `-b` are fundamentally different ways of using Sphinx. However, we want to introduce a new `sphinx` command, so I'm closing in favour of #5618.

A
